src/hubs/generic/basehub.ts

```diff
@@ -317,45 +306,17 @@ export class BaseHub extends EventEmitter {
          * @param {Device} device
          */
         this.emit("detach", device);
-        debug(`Detached device type ${device.type} (${Consts.DeviceTypeNames[device.type]}) on port ${device.portName} (${device.portId})`);
+        debug(`Detached device type ${device.type} (${deviceTypeNames[device.type]}) on port ${device.portName} (${device.portId})`);
     }
 
 
-    protected _createDevice (deviceType: number, portId: number) {
-        let constructor;
-
-        // NK TODO: This needs to go in a better place
-        const deviceConstructors: {[type: number]: typeof Device} = {
-            [Consts.DeviceType.LIGHT]: Light,
-            [Consts.DeviceType.TRAIN_MOTOR]: TrainMotor,
-            [Consts.DeviceType.SIMPLE_MEDIUM_LINEAR_MOTOR]: SimpleMediumLinearMotor,
-            [Consts.DeviceType.MOVE_HUB_MEDIUM_LINEAR_MOTOR]: MoveHubMediumLinearMotor,
-            [Consts.DeviceType.MOTION_SENSOR]: MotionSensor,
-            [Consts.DeviceType.TILT_SENSOR]: TiltSensor,
-            [Consts.DeviceType.MOVE_HUB_TILT_SENSOR]: MoveHubTiltSensor,
-            [Consts.DeviceType.TECHNIC_MEDIUM_HUB_TILT_SENSOR]: TechnicMediumHubTiltSensor,
-            [Consts.DeviceType.TECHNIC_MEDIUM_HUB_GYRO_SENSOR]: TechnicMediumHubGyroSensor,
-            [Consts.DeviceType.TECHNIC_MEDIUM_HUB_ACCELEROMETER]: TechnicMediumHubAccelerometerSensor,
-            [Consts.DeviceType.MEDIUM_LINEAR_MOTOR]: MediumLinearMotor,
-            [Consts.DeviceType.TECHNIC_LARGE_LINEAR_MOTOR]: TechnicLargeLinearMotor,
-            [Consts.DeviceType.TECHNIC_XLARGE_LINEAR_MOTOR]: TechnicXLargeLinearMotor,
-            [Consts.DeviceType.COLOR_DISTANCE_SENSOR]: ColorDistanceSensor,
-            [Consts.DeviceType.VOLTAGE_SENSOR]: VoltageSensor,
-            [Consts.DeviceType.CURRENT_SENSOR]: CurrentSensor,
-            [Consts.DeviceType.REMOTE_CONTROL_BUTTON]: RemoteControlButton,
-            [Consts.DeviceType.HUB_LED]: HubLED,
-            [Consts.DeviceType.DUPLO_TRAIN_BASE_COLOR_SENSOR]: DuploTrainBaseColorSensor,
-            [Consts.DeviceType.DUPLO_TRAIN_BASE_MOTOR]: DuploTrainBaseMotor,
-            [Consts.DeviceType.DUPLO_TRAIN_BASE_SPEAKER]: DuploTrainBaseSpeaker,
-            [Consts.DeviceType.DUPLO_TRAIN_BASE_SPEEDOMETER]: DuploTrainBaseSpeedometer
-        };
-
-        constructor = deviceConstructors[deviceType as Consts.DeviceType];
-        
+    protected _createDevice (deviceType: number, portId: number, versions: DeviceVersion) {
+        const constructor = devices[deviceType];
+
         if (constructor) {
-            return new constructor(this, portId);
+            return new constructor(this, portId, versions);
         } else {
-            return new Device(this, portId, undefined, deviceType);
+            return new Device(this, portId, versions, undefined);
         }
 
     }
```

Comment:
I don't know if it's a better place but now it is in the `devices.ts` collection of devices constructors